File: flutter_app/lib/presentation/wallet/wallet_view_model.dart
import 'package:flutter/material.dart';
import 'package:firebase_auth/firebase_auth.dart' as auth;
import 'package:cloud_firestore/cloud_firestore.dart';

import '../../domain/model/data_models.dart';

class WalletUiState {
  final bool isLoading;
  final double balance;
  final List<WalletTransaction> transactions;
  final String? errorMessage;

  WalletUiState({
    this.isLoading = true,
    this.balance = 0.0,
    this.transactions = const [],
    this.errorMessage,
  });

  WalletUiState copyWith({
    bool? isLoading,
    double? balance,
    List<WalletTransaction>? transactions,
    String? errorMessage,
  }) {
    return WalletUiState(
      isLoading: isLoading ?? this.isLoading,
      balance: balance ?? this.balance,
      transactions: transactions ?? this.transactions,
      errorMessage: errorMessage ?? this.errorMessage,
    );
  }
}

class WalletViewModel extends ChangeNotifier {
  final auth.FirebaseAuth _auth = auth.FirebaseAuth.instance;
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  WalletUiState _uiState = WalletUiState();
  WalletUiState get uiState => _uiState;

  WalletViewModel() {
    fetchWalletData();
  }

  void _setUiState(WalletUiState state) {
    _uiState = state;
    notifyListeners();
  }

  void fetchWalletData() {
    final userId = _auth.currentUser?.uid;
    if (userId == null) return;

    _firestore.collection("users").doc(userId).snapshots().listen((snapshot) {
      if (snapshot.exists) {
        final bal = snapshot.data()?['walletBalance']?.toDouble() ?? 0.0;
        _setUiState(_uiState.copyWith(balance: bal));
      }
    });

    _firestore.collection("transactions")
        .where("userId", isEqualTo: userId)
        .snapshots()
        .listen((snapshot) {
      final list = snapshot.docs.map((doc) => WalletTransaction.fromMap(doc.data())).toList();
      list.sort((a, b) => b.timestamp.compareTo(a.timestamp));
      _setUiState(_uiState.copyWith(isLoading: false, transactions: list));
    }, onError: (e) {
      _setUiState(_uiState.copyWith(isLoading: false, errorMessage: e.toString()));
    });
  }
}

File: app/src/main/java/com/kina/care/presentation/wallet/WalletViewModel.kt
package com.kina.care.presentation.wallet

import androidx.lifecycle.ViewModel
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.kina.care.domain.model.WalletTransaction
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

data class WalletUiState(
    val isLoading: Boolean = true,
    val balance: Double = 0.0,
    val transactions: List<WalletTransaction> = emptyList(),
    val errorMessage: String? = null
)

class WalletViewModel : ViewModel() {
    private val firestore = FirebaseFirestore.getInstance()
    private val auth = FirebaseAuth.getInstance()

    private val _uiState = MutableStateFlow(WalletUiState())
    val uiState: StateFlow<WalletUiState> = _uiState.asStateFlow()

    init {
        fetchWalletData()
    }

    private fun fetchWalletData() {
        val userId = auth.currentUser?.uid ?: return

        // 1. දැනට ඇති මුදල (Balance) ලබාගැනීම
        firestore.collection("users").document(userId)
            .addSnapshotListener { snapshot, error ->
                if (error != null) return@addSnapshotListener
                val balance = snapshot?.getDouble("walletBalance") ?: 0.0
                _uiState.value = _uiState.value.copy(balance = balance)
            }

        // 2. ගනුදෙනු ඉතිහාසය (Transactions) ලබාගැනීම
        firestore.collection("transactions")
            .whereEqualTo("userId", userId)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    _uiState.value = _uiState.value.copy(isLoading = false, errorMessage = error.localizedMessage)
                    return@addSnapshotListener
                }
                if (snapshot != null) {
                    // අලුත්ම ගනුදෙනු ඉහළින් පෙන්වීමට කාලය අනුව පෙළගස්වයි
                    val list = snapshot.documents.mapNotNull { it.toObject(WalletTransaction::class.java) }
                        .sortedByDescending { it.timestamp }
                    _uiState.value = _uiState.value.copy(isLoading = false, transactions = list)
                } else {
                    _uiState.value = _uiState.value.copy(isLoading = false)
                }
            }
    }
}